Add support for ~/nvm/default-packages file

westonganger opened the pull request:
This PR adds support for a file containing default packages to be installed with every new version install. 

While the `reinstall-packages` option is great, this file based approach is nice because I can check this file into my personal dotfiles repo.

Note: this idea was adapted from https://github.com/rbenv/rbenv-default-gems. I wanted a similar thing from my node version manager setup.

ljharb commented on nvm.sh:
```diff
@@ -2404,6 +2404,28 @@ nvm() {
+          DEFAULT_PACKAGES="${DEFAULT_PACKAGES}${line[0]} "
+        done < "${NVM_DIR}/default-packages"
+
+        if [ "$DEFAULT_PACKAGES" != "" ]; then
```
this should probably use `[ -z "${DEFAULT_PACKAGES}" ]`

ljharb commented on nvm.sh:
```diff
@@ -2404,6 +2404,28 @@ nvm() {
+        while IFS=" " read -r -a line; do
+
+          # Skip empty lines.
+          [ "${#line[@]}" -gt 0 ] || continue
```
would `[ -z "${line[@]}" ]` work?

westonganger commented:
I agree we should not go about adding a bunch of unnecessary complexity with this. But from what I have added so far is not very complicated.

For the lines with spaces in them, that should be invalid I assumed... We can easily provide an explanatory error for this but honestly this is something that is along the lines of garbage in, garbage out. The user should simply write a correct default-packages file.

For the node version compatibility with packages specified. Adding this compatibility check, in my opinion, is adding too much complexity to be worth it unless you know of a dead simple way. But regardless, the most common use case for a version manager is upgrading to newer versions, not installing really old versions. For these cases we should provide a `--skip-default-packages` option or alternatively they can simply comment out the offending package from their file before performing an install of an old incompatible version.

I will add the unit tests if you plan on merging this eventually.

westonganger commented:
* I added the changes you requested.
* I have also now added the `--skip-default-packages` option.
* If default-packages contains the space in a line you mentioned before or if the default packages install fails then it will now provide an error to the user telling them to check their `default-packages` files for errors.
* I started adding tests but got a little confused as I havent done this testing before. So its still failings the newly added tests right now. Maybe if someone could help me out a little with the tests.

ljharb commented on README.markdown:
````diff
@@ -216,6 +216,19 @@ nvm install 6 --reinstall-packages-from=5
+If you have a list of default packages you want installed every time you install a new version we support that too. You can add any valid package, version, repo, url for installing an npm package.
+
+```sh
+# ~/.nvm/default-packages
````
let's do `$NVM_DIR/default-packages` here, just to try to avoid hard-coding the default

ljharb commented on README.markdown:
```diff
@@ -216,6 +216,19 @@ nvm install 6 --reinstall-packages-from=5
 
+### Default global packages from file while installing
+
+If you have a list of default packages you want installed every time you install a new version we support that too. You can add any valid package, version, repo, url for installing an npm package.
```
"for installing an npm package - anything npm would accept as a package argument on the command line."?

ljharb commented on nvm.sh:
```diff
@@ -2404,6 +2405,61 @@ nvm() {
+            PROVIDED_REINSTALL_PACKAGES_FROM="$(nvm_echo "$1" | command cut -c 27-)"
+            REINSTALL_PACKAGES_FROM="$(nvm_version "$PROVIDED_REINSTALL_PACKAGES_FROM")" ||:
+          ;;
+          --copy-packages-from=*)
```
i think adding these lines means we can remove the ones on lines ~2467?

ljharb commented on test/fast/Unit tests/nvm_default_packages:
```diff
@@ -0,0 +1,49 @@
+
+# commented-package
+
+webpack/webpack
```
let's pick some smaller packages for the tests (than webpack or gulp) and still-relevant ones (unlike bower) - how about `rimraf`, `object-inspect`, things like that?

ljharb commented on test/fast/Unit tests/nvm_default_packages:
```diff
@@ -0,0 +1,49 @@
+cleanup
+
+cat > ../../../default-packages << EOF
+not-a-package-name
```
this *could* be a valid package name - maybe `not~a~package~name` or something?

ljharb commented on test/fast/Unit tests/nvm_default_packages:
```diff
@@ -0,0 +1,49 @@
+not-a-package-name
+EOF
+
+OUTPUT="$(nvm exec 6.10.1 npm list -g | grep gulp)"
```
`npm ls -g` can take a `--depth=0` argument which should help the test be faster and more accurate

westonganger commented:
I have added the changes you requested.

By the way, Is there a way I can run only the unit tests or only a specific unit test?

ljharb commented on .gitignore:
```diff
@@ -11,6 +11,7 @@ test/**/test_output
 
 node_modules/
 npm-debug.log
+.yarn-lock
```
`yarn.lock` should be gitignored; but I'm not sure what `.yarn-lock` is nor why it would occur in a project not using yarn?

westonganger replied:
Just because I installed the devDependencies using yarn its faster, I thought about this already before adding it to this file. Its just a .gitignore file, let it ignore.

ljharb replied:
Project-level gitignores should only contain project-related ignores; anything else should be in a global gitignore - otherwise you get every IDE's dotfiles polluting every repo.

If you wanted to ignore `yarn.lock` I'd understand, but `.yarn-lock` isn't a file that exists, as far as I'm aware.

ljharb commented on nvm.sh:
```diff
@@ -2419,13 +2421,51 @@ nvm() {
             REINSTALL_PACKAGES_FROM="$(nvm_version "$PROVIDED_REINSTALL_PACKAGES_FROM")" ||:
           ;;
+          --skip-default-packages)
+            SKIP_DEFAULT_PACKAGES=true ||:
```
`||:` shouldn't be required here, since you're not doing anything that could possibly have a nonzero exit code.

ljharb commented on nvm.sh:
```diff
@@ -2419,13 +2421,51 @@ nvm() {
       done
 
+      if [ -z "$SKIP_DEFAULT_PACKAGES" ] && [ -f "${NVM_DIR}/default-packages" ]; then
+        local DEFAULT_PACKAGES=""
```
please always declare `local` vars in a separate line from initializing them; this line as-is won't work in `ksh` which doesn't support `local`.

ljharb commented on nvm.sh:
```diff
@@ -2419,13 +2421,51 @@ nvm() {
+          { 
+            nvm_echo "$DEFAULT_PACKAGES" | command xargs npm install -g --quiet 
+          } || {
+            nvm_echo "Failed installing default packages. Please check if your default-packages file or a package in it has problems!"
```
this is error output; it should use `nvm_err` instead of `nvm_echo`

ljharb commented on test/fast/Unit tests/nvm_default_packages:
```diff
@@ -0,0 +1,49 @@
+
+die () { echo "$@" ; cleanup ; exit 1; }
+cleanup () {
+  rm -rf "$(nvm_version_path v6.10.1)" ../../../default-packages
```
tests are often run inside a working nvm installation; if I have a pre-existing `default-packages` file, this test will clobber it.

ljharb commented:
@westonganger typically i just `cd` into the directory and `./test_file` to run a single one

ljharb commented on nvm.sh:
```diff
@@ -2419,13 +2421,51 @@ nvm() {
+          [ -z "${line}" ] || continue
+
+          # Skip comment lines that begin with `#`.
+          [ "${line:0:1}" != "#" ] || continue
```
the shellcheck job failed; it says string indexing isn't supported in dash.

westonganger commented:
That makes testing much faster. Thanks!

I have now updated this with changes you requested.

ljharb commented on README.markdown:
```diff
@@ -216,6 +217,19 @@ nvm install 6 --reinstall-packages-from=5
+
+gulp
+webpack
+bower@1.7.1
```
let's avoid using bower as an example here too (i can update this myself prior to merging tho, not a blocker)

ljharb commented on nvm.sh:
```diff
@@ -2419,13 +2421,54 @@ nvm() {
 
+      if [ -z "$SKIP_DEFAULT_PACKAGES" ] && [ -f "${NVM_DIR}/default-packages" ]; then
+        local DEFAULT_PACKAGES
+
```
i'll remove this extra blank line also

ljharb commented on test/fast/Unit tests/nvm_default_packages:
```diff
@@ -0,0 +1,74 @@
+
+# commented-package
+
+stevemao/left-pad
```
cute

ljharb commented on .gitignore:
```diff
@@ -11,6 +11,7 @@ test/**/test_output
 node_modules/
 npm-debug.log
+.yarn-lock
 
```
we should probably ignore `default-packages` as well, so contributors don't accidentally commit theirs.

ljharb commented:
Although, tests are still failing - so this will need to wait until they can be made to pass.

westonganger commented:
I added the changes you have mentioned and continued working on the tests. 

I am kind of stuck. Getting the following error: `xargs: npm: No such file or directory`

ljharb commented on nvm.sh:
```diff
@@ -2419,13 +2421,53 @@ nvm() {
+        if [ -z "$DEFAULT_PACKAGES" ]; then
+          nvm_echo "Installing default global packages from ${NVM_DIR}/default-packages..."
+          { 
+            nvm_echo "$DEFAULT_PACKAGES" | command xargs npm install -g --quiet 
```
ah - you're trying to `npm install` before `nvm use` is ran on line 2488.

I think you should extract this block of code to a separate function; and then call that function on line 2495.5.

westonganger commented:
I moved it to another function. Hopefully its what you meant.

ljharb commented:
That looks great; there's other seemingly unrelated tests failing, presumably because they do `nvm install`.

Have you tested locally with `set -e`, to ensure that your new code doesn't cause `nvm` to exit out when there's no default packages, or also when the list installs properly?

westonganger commented:
I have now done all the local testing and came up with some glaring holes, oops... I patched that all up, enhanced and fixed the tests. It now works like a charm under all conditions. I think this one is ready to go!

ljharb commented on nvm.sh:
```diff
@@ -2441,9 +2471,12 @@ nvm() {
 
-      if nvm_is_version_installed "$VERSION"; then
+      if nvm_is_version_installed "$VERSION" && nvm use "$VERSION"; then
         nvm_err "$VERSION is already installed."
```
it's very intentional that the error message prints only if `nvm_is_version_installed` fails, and not if `nvm use` fails.

westonganger replied:
Good call was unsure about that one

ljharb commented on nvm.sh:
```diff
@@ -3257,6 +3293,17 @@ nvm() {
+    nvm_echo "$1" | command xargs npm install -g --quiet && return 0
+  } || {
+    nvm_err "Failed installing default packages. Please check if your default-packages file or a package in it has problems!"
+    return 1
```
maybe this should capture the exit code from `npm install`?

westonganger replied:
Ya this definately had weird style. I have updated it to be more standard.

ljharb commented on nvm.sh:
```diff
@@ -2511,6 +2546,9 @@ nvm() {
+        if [ -z "$SKIP_DEFAULT_PACKAGES" ] && [ -n "$DEFAULT_PACKAGES" ]; then
+          nvm_install_default_packages "$DEFAULT_PACKAGES"
+        fi
         if [ ! -z "$REINSTALL_PACKAGES_FROM" ] \
```
not sure why i used `! -z` here, but this could become `-n` too :-)

westonganger replied:
lol. I looked at that but was like shit it aint broken, better not touch it

westonganger commented:
I removed `set -ex` from those failing tests and they passed. However another different errored test popped up for bash only I can try to do the same for that test too.

Would you merge a different PR which removes all `set -ex` (or maybe just the `x` option) from the tests so that we dont have such a large output to sift through when viewing the test logs.

ljharb commented:
That means the `set -e` is likely what caused the tests to fail; that's a critical part of the tests. the `set -x` is just there so when there *is* a failure, it's easier to debug it. Please do restore those.

westonganger commented:
Alright these are back. I dont know what to do about those tests nothing should have changed regarding those tests.

westonganger commented:
Can this be merged now?

ljharb commented:
@westonganger I'm not sure what to do about the tests either; the implication is that your changes have affected those code paths.

I can't merge it unless tests are passing.

westonganger commented:
So both of these tests run successfully on my local machine, why do they fail on the build?

roberttod commented:
This looks really good! It would be great to manage the packages using nvm rather than editing the file, perhaps after this gets accepted.

ljharb commented:
@westonganger so one of the failures is because your new function `nvm_install_default_packages` isn't being cleared by `nvm unload`. I've fixed that.

ljharb commented:
Thanks for the contribution, and for bearing with the review!

augbog commented:
Thank you for adding this!

ljharb commented:
@Slapbox there isn't a separate command, no - we could certainly add one if there's a use case. file a new issue?
